src/gui/MergeDialog.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ *  Copyright (C) 2017 Weslly Honorato <weslly@protonmail.com>
+ *  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "MergeDialog.h"
+#include "ui_MergeDialog.h"
+
+#include "core/Database.h"
+
+#include <QPushButton>
+#include <QShortcut>
+#include <qdebug.h>
+
+MergeDialog::MergeDialog(QSharedPointer<Database> source, QSharedPointer<Database> target, QWidget* parent)
+    : QDialog(parent)
+    , m_sourceDatabase{std::move(source)}
+    , m_targetDatabase{std::move(target)}
+{
+    setAttribute(Qt::WA_DeleteOnClose);
+
+    m_ui.setupUi(this);
+
+    m_ui.buttonBox->button(QDialogButtonBox::Ok)->setText(tr("Merge"));
+    m_ui.buttonBox->button(QDialogButtonBox::Ok)->setFocus();
+
+    connect(m_ui.buttonBox, SIGNAL(rejected()), SLOT(abortMerge()));
+    connect(m_ui.buttonBox, SIGNAL(accepted()), SLOT(performMerge()));
+
+    setupChangeTable();
+
+    // block input to other windows since other interactions can lead to unexpected merge results
+    setWindowModality(Qt::WindowModality::ApplicationModal);
+}
+
+MergeDialog::MergeDialog(const Merger::ChangeList& changes, QWidget* parent)
+    : QDialog(parent)
+    , m_changes(changes)
+{
+    setAttribute(Qt::WA_DeleteOnClose);
+
+    m_ui.setupUi(this);
+
+    m_ui.buttonBox->button(QDialogButtonBox::Ok)->setFocus();
+    m_ui.buttonBox->button(QDialogButtonBox::Abort)->hide();
+
+    connect(m_ui.buttonBox, SIGNAL(accepted()), SLOT(close()));
+
+    setupChangeTable();
+}
+
+void MergeDialog::setupChangeTable()
+{
+    if (m_changes.isEmpty()) {
+        // deep copy of root group with same uuid
+        auto tmpRootGroup = m_targetDatabase->rootGroup()->clone(Entry::CloneFlag::CloneIncludeHistory,
+                                                                 Group::CloneFlag::CloneIncludeEntries);
+        Database tmpDatabase;
+        tmpDatabase.setRootGroup(tmpRootGroup);
+        m_changes = Merger(m_sourceDatabase.data(), &tmpDatabase).merge();
```

Comment:
@droidmonkey while this kind works, is there a better way to clone a database? Wasn't able to find one in the current code base.
Also, this does not include the `Database::m_metadata` which doesn't leads to missing changes regarding Recycling Bin and Custom Icons - is there a reason why it doesn't implement a `clone()` function or should I implement it to do a full deep copy of a database?

Reply:
I'd have to check but I don't think so since cloning a database is not really a thing in the program. Not saying it wouldn't be useful, but the current method you are using here is not the correct way. That would be more like transferring root group and children.

Reply:
But doesn't `Group::clone()` create an independent copy of the root group (the ownership of which I then transfer to the temporary database)?

The alternative to cloning the database would be to implement something like `discardModifications()` for a database - as far as I can see, that doesn't exist yet. But then we have to be absolutely sure that no one will save the database in the meantime before the merge is confirmed.

Which of these two approaches do you think fit the current design better? Or do you have another idea on how to achieve such a "temporary merge" without major changes to the merge logic?